bip-tx-terminology.mediawiki

```diff
@@ -0,0 +1,324 @@
+: Synonyms: <code>scriptSig</code>, unlocking script (see [[locking_unlocking|Rationale]])
+
+; leaf script
+: [Component] A condition script used for spending P2TR outputs via the scriptpath. The leaf script appears as a witness item.
+
+; leaf version
+: [Component] The version of Script used in a leaf script. At the time of writing, only Tapscript is defined.
+
+; lock time
+: [Component] Part of the transaction header. Lock time specifies a time (in blockheight or UNIX time) before which a transaction cannot be included in a block if any input’s sequence number indicates non-finality. The lock time can also indicate a relative spending lock in combination with <code>OP_CHECKSEQUENCEVERIFY</code>. Serialized as the four-byte <code>nLockTime</code> integer field.
+: Synonym: <code>nLockTime</code>
+
+; marker
+: [Artifact] 1-byte serialization artifact indicating that a type of extended serialization is being used for this transaction. Must always be <code>00</code> (see [[https://github.com/bitcoin/bips/blob/master/bip-0144.mediawiki|BIP144]]). (Note: Non-segwit nodes will only accept stripped segwit transactions, because the marker appears at the position where non-segwit nodes expect the input counter. The input counter may not be zero, so a complete (non-stripped) segwit transaction appears invalid to a non-segwit node.)
+
+; outpoint
+: [Component] Identifies the transaction output (TXO) being spent by a transaction input. Consists of a txid and output index. The txid is serialized in little-endian but displayed in big-endian.
+
+; output counter
+: [Artifact] The length of the transaction output list, (see [[https://github.com/bitcoin/bips/blob/master/bip-0144.mediawiki|BIP144]]).
+: Synonym: <code>txout_count</code>
+
+; output index
+: [Component] Part of an outpoint. The position of the output in a transaction’s output list that created the identified TXO.
+
+; output script
+: [Component] Must be present in each transaction output. Contains either a condition script or a forwarding script. Originally, the Bitcoin code base used Hungarian notation. This field was presumably named <code>scriptPubKey</code> to refer to the transaction field of the type script that contained the public key. The use of this field had evolved beyond that interpretation even when Bitcoin was published since the field could already contain more complicated scripts. Today, this field is referred to as the output script.
```

Comment:
The "presumably" bit doesn't seem right to me? I would have assumed the thinking was "okay, so we have an elliptic-curve public key to decide who can spend it, and then an ECDSA signature from that public key to actually spend it" "yeah, but what about generalising it for multisig or only-if-you-win-this-hand-of-poker or whatever?" "oh, cool cool cool. so let's have a scripting language so you can specify arbitrary conditions?" "and then we'll just turn both the pubkey and signature into scripts -- the scriptPubKey will be a public script that only the people allowed to spend can satisfy, and they'll do that with a script of their own which we'll call the scriptSig" "nice! you're so smart satoshi!" "thanks satoshi!".

That is, "scriptPubKey" is a script (that's it type) and it's also a "PubKey" (that's its purpose -- it's the public part that everyone can know, for which there's a corresponding private part that someone can use to generate a signature), it's not that the "scriptPubKey" is a script that happens to contain ECC pubkeys. That got changed with p2sh and segwit, where additional implicit constraints beyond what is specified in the script are imposed; but in that case "output script" isn't really an accurate description either for the same reason: p2sh, p2wpkh, p2wsh and p2tr scriptPubKeys aren't really scripts in any meaningful sense.

I'm not really sure it even makes sense to come up with a single term to describe all the current scriptPubKey possibilities -- p2pk, p2pkh, p2ms, p2anchor are all scripts, but p2sh, p2wsh are just pushing hashes of scripts, p2wpkh is a push of a pubkey hash, and p2tr pushes a direct pubkey that may also involve many hidden script hashes. Not all of those involve a "script", not all of them involve an ECC pubkey, p2anchor doesn't even put any constraints on spending. Maybe "spending constraint commitment" would be accurate? (The constraint itself is "a signature by this pubkey" or "give me a script that hashes to this, and then an initial stack for the script that results in successful execution")

Reply:
I like your interpretation of the origin of `scriptPubKey`, perhaps it could be included like this? (I'm sure someone can write it more elegantly)

```suggestion
: [Component] Must be present in each transaction output. Contains either a condition script or a forwarding script. Originally, the Bitcoin code base used Hungarian notation. This field was presumably named <code>scriptPubKey</code> as it was of the type script and either because it contained the public key or because it fulfilled the role of a public key. The use of this field had evolved beyond that interpretation even when Bitcoin was published since the field could already contain more complicated scripts. Today, this field is referred to as the output script.
```

However, "output script" is, in my opinion, by far the best name for this field, even for scripts with implied meaning ("forwarding scripts"), unless someone can convince me that this shouldn't be called a script:
```
OP_0 OP_PUSHBYTES_20 7de8b69f55bebe02f99f77b2c8f9848b915f1953
```